### PR TITLE
Add IPFabric integration

### LINKED ISSUE #129: Migrate Cisco ACI SSoT Plugin to This Repo
### Environment
* Nautobot version:  
* nautobot-plugin-chatops version:  

### Proposed Functionality

After #128 and #115, merge https://github.com/nautobot/nautobot-plugin-ssot-aci into this repo as an `integration`.

### Use Case

Make installation, discovery, maintaining easier.

### LINKED ISSUE #130: Migrate IP Fabric SSoT Plugin to This Repo
### Environment
* Nautobot version:  
* nautobot-plugin-chatops version:  

### Proposed Functionality

After #128 and #115, merge https://github.com/nautobot/nautobot-plugin-ssot-ipfabric into this repo as an `integration`.

### Use Case

Make installation, discovery, maintaining easier.

### PR DESCRIPTION
**Pull request by snaselj**
# Closes #130 

- Adds IPFabric integration from nautobot-plugin-ssot-ipfabric.

## TODO

- [x] Test integration.
- [x] Fix configuration.
- [x] Bump dependency to the IPFabric integraton version of `nautobot-chatops` App.

### CODE REVIEW

File: development/nautobot_config.py [hunk 1]
```diff
@@ -132,13 +132,28 @@
 #
 
 # Enable installed plugins. Add the name of each plugin to the list.
-PLUGINS = ["nautobot_ssot"]
+PLUGINS = [
+    "nautobot_chatops",
+    "nautobot_chatops_ipfabric",
```
Comment (snaselj) [inline development/nautobot_config.py:137]: Update after nautobot-chatops with ipfabric integration will be deployed.

File: development/nautobot_config.py [hunk 2]
```diff
@@ -132,13 +132,28 @@
...
+    "nautobot_chatops": {
+        "enable_slack": True,
+        "slack_api_token": os.getenv("SLACK_API_TOKEN"),
+        "slack_signing_secret": os.getenv("SLACK_SIGNING_SECRET"),
+        "session_cache_timeout": 3600,
+    },
+    "nautobot_chatops_ipfabric": {
+        "IPFABRIC_API_TOKEN": os.getenv("IPFABRIC_API_TOKEN"),
+        "IPFABRIC_HOST": os.getenv("IPFABRIC_HOST"),
+    },
```
Comment (snaselj) [inline development/nautobot_config.py:153]: Update after nautobot-chatops with ipfabric integration will be deployed.

File: pyproject.toml [hunk 3]
```diff
@@ -61,20 +66,38 @@ mkdocs-version-annotations = "1.0.0"
 mkdocstrings = "0.20"
 mkdocstrings-python = "0.8.3"
 requests-mock = "^1.10.0"
+responses = "^0.14.0"
+Sphinx = "^4.3.1"
+myst-parser = "^0.15.2"
+sphinx-autoapi = "^1.8.4"
+sphinx-rtd-theme = "^1.0.0"
+nautobot-chatops = "^1.5.1"
+nautobot-chatops-ipfabric = "^3.0.0"
```
Comment (snaselj) [inline pyproject.toml:75]: Update after nautobot-chatops with ipfabric integration will be deployed.

File: nautobot_ssot/integrations/ipfabric/diffsync/adapter_ipfabric.py [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+from nautobot_ssot.integrations.ipfabric.diffsync import DiffSyncModelAdapters
+
+logger = logging.getLogger("nautobot.jobs")
+
+CONFIG = settings.PLUGINS_CONFIG.get("nautobot_ssot", {})
+DEFAULT_INTERFACE_TYPE = CONFIG.get("default_interface_type", "1000base-t")
+DEFAULT_INTERFACE_MTU = CONFIG.get("default_interface_mtu", 1500)
+DEFAULT_INTERFACE_MAC = CONFIG.get("default_interface_mac", "00:00:00:00:00:01")
+DEFAULT_DEVICE_ROLE = CONFIG.get("default_device_role", "Network Device")
+DEFAULT_DEVICE_STATUS = CONFIG.get("default_device_status", "Active")
```
Comment (snaselj) [inline nautobot_ssot/integrations/ipfabric/diffsync/adapter_ipfabric.py:21]: These config keys are not present in nautobot_config.py neither in PluginConfig default keys.

File: nautobot_ssot/integrations/ipfabric/jobs.py [hunk 5]
```diff
@@ -0,0 +1,283 @@
...
+        return {
+            "IP Fabric host": CONFIG["ipfabric_host"],
+            "Default MAC Address": CONFIG.get("default_interface_mac", "00:00:00:00:00:01"),
+            "Default Device Role": CONFIG.get("default_device_role", "Network Device"),
+            "Default Interface Type": CONFIG.get("default_interface_type", "1000base-t"),
+            "Default Device Status": CONFIG.get("default_device_status", "Active"),
+            "Allow Duplicate Addresses": CONFIG.get("allow_duplicate_addresses", True),
+            "Default MTU": CONFIG.get("default_interface_mtu", 1500),
+            "Nautobot Host URL": CONFIG.get("nautobot_host"),
+            "Safe Delete Device Status": CONFIG.get("safe_delete_device_status", "Deprecated"),
```
Comment (snaselj) [inline nautobot_ssot/integrations/ipfabric/jobs.py:203]: Different default values

**Comment by snaselj** [comment]
Merged with the latest develop to resolve conflicts

File: pyproject.toml [hunk 6]
```diff
@@ -81,7 +81,7 @@ mkdocstrings-python = "0.8.3"
 requests-mock = "^1.10.0"
 parameterized = "^0.8.1"
 myst-parser = "^0.15.2"
-nautobot-chatops = "^2.0.0"
+nautobot-chatops = { version = "^2.0.2", extras = ["ipfabric"] }
```
Comment (bryanculver) [inline pyproject.toml:84]: 2.0.2 doesn't exist yet. 2.0.1 was just released. Please update the lockfile after this change.

**Comment by snaselj** [comment]
Blocked by https://github.com/nautobot/nautobot-plugin-ssot/pull/157

File: pyproject.toml (hunk 6 above)
Comment (snaselj) [reply pyproject.toml:84]: Fixed

**Review by chadell — COMMENTED** [review]
I think that the integration makes sense. I would only ask @alhogan or @pke11y to test it and verify that it works as expected

**Comment by alhogan** [comment]
I will carve out some time to test it this week.

**Comment by alhogan-aa** [comment]
I ran a few tests and saw some discrepancies in the number of objects been created/updated/deleted. I will dig a little deeper into the results to see what is causing this and provide an update shortly.

**Comment by alhogan** [comment]
I ran a few tests and saw some discrepancies in the number of objects been created/updated/deleted. I will dig a little deeper into the results to see what is causing this and provide an update shortly.

**Comment by alhogan** [comment]
Found the source of the discrepancies. Null interface descriptions were throwing false-positives in the diff. Fix has been implemented and we should be good to go on merging this one.
Screen Shot 2023-08-31 at 6 50 21 PM

**Pull request merged**